In [1]:
import pandas as pd
import sqlite3

# Extract, Transform, Load (ETL)

## Customers Dataset

In this step, the customers dataset is loaded from an external source in CSV format, hosted on a GitHub repository. The data is imported using the read_csv() function from the pandas library and stored in the variable customers_df. To gain an initial understanding of the structure and contents of the dataset, the first five rows are displayed using the head() function.

In [2]:
customers_url = "https://raw.githubusercontent.com/aryashidika/Brazilian-Store-Analysis/refs/heads/main/data/raw/olist_customers_dataset.csv"
customers_df = pd.read_csv(customers_url)
customers_df.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

The info() method is used to display a concise summary of the DataFrame. It provides important information such as the number of entries, column names, data types, and the number of non-null values in each column. This helps to quickly assess the completeness of the data and identify potential missing values or data type issues that may need to be addressed during preprocessing.

In [3]:
customers_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


It can be seen that there are no columns with null or missing values, and the data types are appropriate.

The next line checks for duplicate entries in the customer_id column of the customers_df DataFrame. It returns the total number of duplicate customer IDs, helping to ensure data uniqueness.

In [4]:
customers_df.duplicated(subset=['customer_id']).sum()

0

This block of code removes leading and trailing whitespace characters from key text columns in the customers_df DataFrame: customer_id, customer_unique_id, customer_city, and customer_state. This is done using the apply() function with a lambda expression, ensuring consistent formatting and preventing potential issues during analysis or merging operations.

In [5]:
customers_df['customer_id'] = customers_df['customer_id'].apply(lambda x: str(x).strip())
customers_df['customer_unique_id'] = customers_df['customer_unique_id'].apply(lambda x: str(x).strip())
customers_df['customer_city'] = customers_df['customer_city'].apply(lambda x: str(x).strip())
customers_df['customer_state'] = customers_df['customer_state'].apply(lambda x: str(x).strip())

## Orders Dataset

The next data will also be loaded using the pandas library and stored in the variable orders_df, with the first five rows displayed to gain an initial understanding of the structure and contents of the dataset.

In [6]:
orders_url = "https://raw.githubusercontent.com/aryashidika/Brazilian-Store-Analysis/refs/heads/main/data/raw/olist_orders_dataset.csv"
orders_df = pd.read_csv(orders_url)
orders_df.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

The following line removes the 'order_approved_at' and 'order_delivered_carrier_date' columns from the orders_df DataFrame. These columns are excluded as they are not required for the subsequent analysis.

In [7]:
orders_df = orders_df.drop(columns=['order_approved_at', 'order_delivered_carrier_date'])

The next step will examine the summary of the data using info(), which provides information on missing values and data types.

In [8]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 6 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_delivered_customer_date  96476 non-null  object
 5   order_estimated_delivery_date  99441 non-null  object
dtypes: object(6)
memory usage: 4.6+ MB


In the summary above, it is observed that the order_delivered_customer_date column contains some null values, and several other columns do not have the correct data types.

Next, we will examine how many duplicates exist in the data, specifically in the customer_id and order_id columns.

In [9]:
print(orders_df.duplicated(subset=['customer_id']).sum())
print(orders_df.duplicated(subset=['order_id']).sum())

0
0


The following code will check how many null values are present in each column.

In [10]:
orders_df.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

The order_delivered_customer_date column contains 2,965 missing values. Further investigation will be conducted by examining the order_status to understand the context of these missing values.

In [11]:
print(orders_df[orders_df['order_delivered_customer_date'].isna()]['order_status'].value_counts())

order_status
shipped        1107
canceled        619
unavailable     609
invoiced        314
processing      301
delivered         8
created           5
approved          2
Name: count, dtype: int64


The following code will identify orders with a status of 'delivered' but without a delivery date. These orders will then be removed from the dataset.

In [12]:
orders_df = orders_df[~((orders_df['order_status'] == 'delivered') & (orders_df['order_delivered_customer_date'].isna()))]

The following line of code checks whether there are any records with a non-null delivery date (order_delivered_customer_date) but a status that is not marked as "delivered". This helps identify potential inconsistencies in the delivery information.

In [13]:
invalid_rows = orders_df.loc[(orders_df['order_status'] != 'delivered') & orders_df['order_delivered_customer_date'].notna()]
invalid_rows['order_status'].unique()

array(['canceled'], dtype=object)

After identifying the status of the invalid rows, the next step is to remove them from the dataset. Dan semua yang memiliki missing value di kolom 'order_delivered_customer_date' dibiarkan karena sudah sesuai tidak mempunyai tanggal delivered.

In [14]:
orders_df = orders_df[~((orders_df['order_status'] == 'canceled') & orders_df['order_delivered_customer_date'].notna())]

The following lines convert the order_purchase_timestamp, order_delivered_customer_date, and order_estimated_delivery_date columns to datetime format using pd.to_datetime(). This ensures that the date-related data can be properly interpreted and analyzed.

In [15]:
orders_df['order_purchase_timestamp'] = pd.to_datetime(orders_df['order_purchase_timestamp'])
orders_df['order_delivered_customer_date'] = pd.to_datetime(orders_df['order_delivered_customer_date'])
orders_df['order_estimated_delivery_date'] = pd.to_datetime(orders_df['order_estimated_delivery_date'])

The following lines apply string cleaning to the order_id, customer_id, and order_status columns. Each value is converted to a string and stripped of leading and trailing whitespace using the strip() method. This step helps ensure consistency in textual data for accurate analysis and merging.

In [16]:
orders_df['order_id'] = orders_df['order_id'].apply(lambda x: str(x).strip())
orders_df['customer_id'] = orders_df['customer_id'].apply(lambda x: str(x).strip())
orders_df['order_status'] = orders_df['order_status'].apply(lambda x: str(x).strip())

## Order Items Dataset

The next data will also be loaded using the pandas library and stored in the variable order_items_df, with the first five rows displayed to gain an initial understanding of the structure and contents of the dataset.

In [17]:
order_items_url = "https://raw.githubusercontent.com/aryashidika/Brazilian-Store-Analysis/refs/heads/main/data/raw/olist_order_items_dataset.csv"
order_items_df = pd.read_csv(order_items_url)
order_items_df.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

The following line removes the 'seller_id' and 'shipping_limit_date' columns from the order_items_df DataFrame. These columns are excluded as they are not required for the subsequent analysis.

In [18]:
order_items_df = order_items_df.drop(columns=['seller_id', 'shipping_limit_date'])

The next step will examine the summary of the data using info(), which provides information on missing values and data types.

In [19]:
order_items_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 5 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   order_id       112650 non-null  object 
 1   order_item_id  112650 non-null  int64  
 2   product_id     112650 non-null  object 
 3   price          112650 non-null  float64
 4   freight_value  112650 non-null  float64
dtypes: float64(2), int64(1), object(2)
memory usage: 4.3+ MB


It can be observed that the dataset does not contain any null values.

Next, the number of duplicate values in the 'order_id' column and duplicate rows in the dataset will be examined.

In [20]:
print("Duplicated order_id: ", order_items_df.duplicated(subset=['order_id']).sum())
print("Duplicated rows: ", order_items_df.duplicated().sum())

Duplicated order_id:  13984
Duplicated rows:  0


The order_id column contains 13,984 duplicate entries. The next step is to examine the structure of the rows with duplicated order_id values.

In [21]:
df_duplicates = order_items_df[order_items_df.duplicated(subset=['order_id'], keep=False)]
df_duplicates.head(10)

order_id  order_item_id  \
13  0008288aa423d2a3f00fcb17cd7d8719              1   
14  0008288aa423d2a3f00fcb17cd7d8719              2   
32  00143d0f86d6fbd9f9b38ab440ac16f5              1   
33  00143d0f86d6fbd9f9b38ab440ac16f5              2   
34  00143d0f86d6fbd9f9b38ab440ac16f5              3   
42  001ab0a7578dd66cd4b0a71f5b6e1e41              1   
43  001ab0a7578dd66cd4b0a71f5b6e1e41              2   
44  001ab0a7578dd66cd4b0a71f5b6e1e41              3   
48  001d8f0e34a38c37f7dba2a37d4eba8b              1   
49  001d8f0e34a38c37f7dba2a37d4eba8b              2   

                          product_id  price  freight_value  
13  368c6c730842d78016ad823897a372db  49.90          13.37  
14  368c6c730842d78016ad823897a372db  49.90          13.37  
32  e95ee6822b66ac6058e2e4aff656071a  21.33          15.10  
33  e95ee6822b66ac6058e2e4aff656071a  21.33          15.10  
34  e95ee6822b66ac6058e2e4aff656071a  21.33          15.10  
42  0b0172eb0fd18479d29c3bc122c058c2  24.89          17.63  
43  0b0172eb0fd18479d29c3bc122c058c2  24.89          17.63  
44  0b0172eb0fd18479d29c3bc122c058c2  24.89          17.63  
48  e67307ff0f15ade43fcb6e670be7a74c  18.99           7.78  
49  e67307ff0f15ade43fcb6e670be7a74c  18.99           7.78

As observed from the table, the order_id values appear duplicated because some orders include more than one product. Each product purchased within the same order is listed separately, as indicated by the order_item_id column.

The following lines apply the strip() method to the order_id and product_id columns of the order_items_df DataFrame. This ensures that any leading or trailing spaces in the values are removed, and the values are converted to strings using str() for consistent formatting.

In [22]:
order_items_df['order_id'] = order_items_df['order_id'].apply(lambda x: str(x).strip())
order_items_df['product_id'] = order_items_df['product_id'].apply(lambda x: str(x).strip())

## Products Dataset

The next data will also be loaded using the pandas library and stored in the variable products_df, with the first five rows displayed to gain an initial understanding of the structure and contents of the dataset.

In [23]:
products_url = "https://raw.githubusercontent.com/aryashidika/Brazilian-Store-Analysis/refs/heads/main/data/raw/olist_products_dataset.csv"
products_df = pd.read_csv(products_url)
products_df.tail()

product_id              product_category_name  \
32946  a0b7d5a992ccda646f2d34e418fff5a0                   moveis_decoracao   
32947  bf4538d88321d0fd4412a93c974510e6  construcao_ferramentas_iluminacao   
32948  9a7c6041fa9592d9d9ef6cfe62a71f8c                    cama_mesa_banho   
32949  83808703fc0706a22e264b9d75f04a2e             informatica_acessorios   
32950  106392145fca363410d287a815be6de4                    cama_mesa_banho   

       product_name_lenght  product_description_lenght  product_photos_qty  \
32946                 45.0                        67.0                 2.0   
32947                 41.0                       971.0                 1.0   
32948                 50.0                       799.0                 1.0   
32949                 60.0                       156.0                 2.0   
32950                 58.0                       309.0                 1.0   

       product_weight_g  product_length_cm  product_height_cm  \
32946           12300.0               40.0               40.0   
32947            1700.0               16.0               19.0   
32948            1400.0               27.0                7.0   
32949             700.0               31.0               13.0   
32950            2083.0               12.0                2.0   

       product_width_cm  
32946              40.0  
32947              16.0  
32948              27.0  
32949              20.0  
32950               7.0

The following line removes the 'product_name_lenght', 'product_description_lenght', 'product_photos_qty', 'product_length_cm', 'product_height_cm', and 'product_width_cm' columns from the products_df DataFrame. These columns are excluded as they are not required for the subsequent analysis.

In [24]:
products_df = products_df.drop(columns=['product_weight_g','product_name_lenght', 'product_description_lenght', 'product_photos_qty', 'product_length_cm', 'product_height_cm', 'product_width_cm'])

The next step will examine the summary of the data using info(), which provides information on missing values and data types.

In [25]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 2 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   product_id             32951 non-null  object
 1   product_category_name  32341 non-null  object
dtypes: object(2)
memory usage: 515.0+ KB


In the summary above, it is observed that the 'product_category_name' column contains some null values, and several other columns do not have the correct data types.

Next, we will examine how many duplicates exist in the data.

In [26]:
print("Duplicated rows: ", products_df.duplicated().sum())

Duplicated rows:  0


In [27]:
print("Duplicated product id: ", products_df.duplicated(subset=['product_id']).sum())

Duplicated product id:  0


The next step is to check the number of missing values in each column.

In [28]:
products_df.isna().sum()

product_id                 0
product_category_name    610
dtype: int64

There are 610 missing values in the product_category_name column. These missing values will be replaced with the value "Unknown".

In [29]:
products_df['product_category_name'].fillna("Unknown", inplace=True)

The following lines of code remove any leading or trailing whitespace from the product_id and product_category_name columns in the products_df DataFrame. The apply() function is used with a lambda function that converts each value to a string and applies the strip() method to eliminate any extra spaces.

In [30]:
products_df['product_id'] = products_df['product_id'].apply(lambda x: str(x).strip())
products_df['product_category_name'] = products_df['product_category_name'].apply(lambda x: str(x).strip())

## Geolocation Dataset

The next data will also be loaded using the pandas library and stored in the variable geolocation_df, with the first five rows displayed to gain an initial understanding of the structure and contents of the dataset.

In [31]:
geolocation_url = "https://raw.githubusercontent.com/aryashidika/Brazilian-Store-Analysis/refs/heads/main/data/raw/olist_geolocation_dataset.csv"
geolocation_df = pd.read_csv(geolocation_url)
geolocation_df.head()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037       -23.545621       -46.639292   
1                         1046       -23.546081       -46.644820   
2                         1046       -23.546129       -46.642951   
3                         1041       -23.544392       -46.639499   
4                         1035       -23.541578       -46.641607   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP

The following line removes the 'geolocation_city', 'geolocation_lat', and 'geolocation_lng' columns from the orders_df DataFrame. These columns are excluded as they are not required for the subsequent analysis.

In [32]:
geolocation_df = geolocation_df.drop(columns=['geolocation_city','geolocation_lat','geolocation_lng'])

The next step will examine the summary of the data using info(), which provides information on missing values and data types.

In [33]:
geolocation_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 2 columns):
 #   Column                       Non-Null Count    Dtype 
---  ------                       --------------    ----- 
 0   geolocation_zip_code_prefix  1000163 non-null  int64 
 1   geolocation_state            1000163 non-null  object
dtypes: int64(1), object(1)
memory usage: 15.3+ MB


It can be observed that the dataset does not contain any null values.

Next, the number of duplicate values in the dataset will be examined.

In [34]:
print("Duplicated row: ", geolocation_df.duplicated().sum())

Duplicated row:  981140


In [35]:
geolocation_duplicated = geolocation_df.duplicated()
geolocation_df[geolocation_duplicated].head(5)

geolocation_zip_code_prefix geolocation_state
2                          1046                SP
10                         1013                SP
13                         1012                SP
14                         1037                SP
15                         1046                SP

It can be observed that the rows are truly duplicated. Therefore, the duplicated rows will be removed in the next step.

In [36]:
geolocation_df.drop_duplicates(inplace=True)

Next, the 'geolocation_state' column will be standardized by removing any leading or trailing whitespace to ensure consistency in the data.

In [37]:
geolocation_df['geolocation_state'] = geolocation_df['geolocation_state'].apply(lambda x: str(x).strip())

## Category Name Translation

The next data will also be loaded using the pandas library and stored in the variable category_translation_df, with the first five rows displayed to gain an initial understanding of the structure and contents of the dataset.

In [38]:
category_translation_url = "https://raw.githubusercontent.com/aryashidika/Brazilian-Store-Analysis/refs/heads/main/data/raw/product_category_name_translation.csv"
category_translation_df = pd.read_csv(category_translation_url)
category_translation_df.head()

product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto
3         cama_mesa_banho                bed_bath_table
4        moveis_decoracao               furniture_decor

The next step will examine the summary of the data using info(), which provides information on missing values and data types.

In [39]:
category_translation_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   product_category_name          71 non-null     object
 1   product_category_name_english  71 non-null     object
dtypes: object(2)
memory usage: 1.2+ KB


It can be seen that there are no columns with null or missing values, and the data types are appropriate.

The next line checks for duplicate entries in the customer_id column of the category_translation_df DataFrame.

In [40]:
print("Duplicated_rows: ", category_translation_df.duplicated().sum())

Duplicated_rows:  0


In this step, both the product_category_name and product_category_name_english columns in the category_translation_df DataFrame are standardized by removing any leading or trailing whitespace. This ensures consistency and avoids potential mismatches due to unintended spaces.

In [41]:
category_translation_df['product_category_name'] = category_translation_df['product_category_name'].apply(lambda x: str(x).strip())
category_translation_df['product_category_name_english'] = category_translation_df['product_category_name_english'].apply(lambda x: str(x).strip())

## Load to Database

In this step, a connection to the SQLite database is established using the sqlite3.connect() function. The database file is located at the specified path (../database/cleaned_data.db) and the connection is stored in the variable conn for use in subsequent database operations.

In [42]:
db_directory = "../database/cleaned_data.db"
conn = sqlite3.connect(db_directory)

In this step, several cleaned DataFrames are written to the SQLite database as separate tables using the to_sql() method. Each DataFrame is saved under a specified table name, with if_exists="replace" ensuring that any existing tables with the same name are replaced. The index=False parameter prevents pandas from writing the DataFrame index as a separate column.

The following tables are created:
- customers
- orders
- order_items
- products
- geolocation
- cat_translate

This process prepares the data for efficient querying and analysis using SQL.

In [43]:
customers_df.to_sql("customers", conn, if_exists="replace", index=False)
orders_df.to_sql("orders", conn, if_exists="replace", index=False)
order_items_df.to_sql("order_items", conn, if_exists="replace", index=False)
products_df.to_sql("products", conn, if_exists="replace", index=False)
geolocation_df.to_sql("geolocation", conn, if_exists="replace", index=False)
category_translation_df.to_sql("cat_translate", conn, if_exists="replace", index=False)

71

The conn.close() command is used to properly close the connection to the SQLite database. This is a good practice to release resources and ensure that all transactions are finalized.

In [44]:
conn.close()